text/0000-machine-input.md

```diff
@@ -0,0 +1,183 @@
+- **Start Date:** 2022-05-23
+- **RFC PR:** (leave this empty)
+
+# Machine Input
+
+## Summary
+
+Input data is a very important concept in server-side workflows, and very useful for state machines + statecharts in general. The general idea is that a workflow (represented by a state machine) starts with input data, which can be from some sort of trigger. That data is then used to initialize the workflow with necessary contextual data, or the data itself may be transformed through the workflow.
+
+This RFC proposes to add input data to the built-in `{ type: 'xstate.init' }` event object in XState, in order to avoid the awkward (and potentially unsafe) workarounds that currently need to be done to support this, and to enable better visualization for representing this input data semantically.
+
+## Motivation
+
+Currently, there are 3 ways to implicitly "provide" input data to an XState machine:
+
+**Overriding `context`**
+
+```js
+const greetMachine = createMachine({
+  context: {
+    name: '' // Need to specify "empty" values, which is unintuitive
+  },
+  entry: (ctx) => { console.log(`Hello, ${ctx.name}!`); }
+});
+
+// Need to overwrite context with input values
+// WARNING: in v4, this will overwrite context fully, not partially
+interpret(greetMachine.withContext({
+  name: 'David'
+})).start();
+```
+
+Problems:
+- Empty values need to be specified, since there is an _implicit_ expectation that these context values are available
+- No separation between "private" context and "public" context meant _only_ for input
+- Too easy to overwrite context values not intended to be overwritten
+- Context values may be invalid, especially when machine interpreted in dynamic environments where type checking is not available/forgotten
+- No indication of which properties are input properties
+- Input values may not need to be long-lived after machine start, yet they reside in `context` persistently
+
+**Machine factory**
+
+```js
+const createGreetMachine = (name) => createMachine({
+  entry: (ctx) => { console.log(`Hello, ${ctx.name}!`); }
+});
+
+interpret(createGreetMachine('David')).start();
+```
+
+Problems:
+- Machine definition is oblivious to any input properties
+- Machine is dynamically created; cannot guarantee that the machine is the same shape every time (even though it should always be)
```

Comment:
Is this unique to this factory approach though? The shape of the definition is dictated by the `createMachine` argument's shape and not based on the position of this `createMachine` call in the file.

Reply:
It is, yes. I'm revising the example to show this.